### DISCUSSION TITLE
Definition of Joint Jacobian

### DISCUSSION THREAD

**Discussion by LeoMaglanoc** [stack-of-tasks/pinocchio#1837] [Q&A]
Hello,

there is a mismatch between my understanding of what a joint jacobian is and what actually gets computed as a joint jacobian in Pinocchio. My goal is to compute the linear and angular velocity at each joint (and also for the end-effector). Is that something I can achieve with the joint jacobian? And if not, are there other Pinocchio functions I could use to do that?

E.g., given the following (pseudo-)code:

```
Eigen::Vector<double, 6> q; 
q << 1.0, 2.0, 3.0, 4.0, 5.0, 6.0;
Eigen::Vector<double, 6> q_dot; 
q_dot << 1.0, 2.0, 3.0, 4.0, 5.0, 6.0;
pinocchio::forwardKinematics(model, data, q);
pinocchio::computeJointJacobians(model, data, q);
pinocchio::getJointJacobian(model, data, joint_id, pinocchio::WORLD, jacobian);
return jacobian * q_dot;
```

Would this return the vector consisting of the linear and angular velocity at joint joint_id?

**Comment by stephane-caron** [comment]
AFAIU yes. I guess you've already checked the documentation for [`getJointJacobian`](https://gepettoweb.laas.fr/doc/stack-of-tasks/pinocchio/master/doxygen-html/namespacepinocchio.html#abadfcd21d13fa17865a189e4b47acca9).

Here is a proposal (:warning: to be validated by project maintainers!) to recap a bit more details about this Jacobian: https://github.com/stack-of-tasks/pinocchio/pull/1836/files

**Comment by jcarpent** [comment]
@LeoMaglanoc You should change the frame of expression (`pinocchio::LOCAL` for getting the joint Jacobian expressed in the LOCAL frame of the joint or `pinocchio::LOCAL_WORLD_ALIGNED` to compute the Jacobian of the joint but expressed in the coordinate system of the WORLD).

**Comment by LeoMaglanoc** [comment]
@jcarpent Thanks, `pinocchio::LOCAL_WORLD_ALIGNED` is probably what I'm looking for. Just for my understanding though, what exactly do you get if you use `pinocchio::WORLD`?

**Comment by jcarpent** [comment]
It will be the velocity of the Jacobian of the point coincident with the origin of the WORLD and expressed in the WORLD coordinate system. I do think this cheat sheet will be helpful: https://github.com/stack-of-tasks/pinocchio/blob/master/doc/pinocchio_cheat_sheet.pdf

**Comment by jcarpent** [comment]
I will close this issue and transfer this issue as Q&A in the discussions.

**Comment by edward9503** [comment]
Hi @jcarpent ,

after reading your discussion and also other issues in github, I still think WORLD is the correct answer. But why LOCAL_WORLD_ALIGNED is always recommanded to get the linear and angular velocity? Maybe I didn't fully understand the difference between WORLD and LOCAL_WORLD_ALIGNED. Thanks.
